feat(api): expand emulators and integration tests.

amitlissack opened the pull request:
# Overview

As I work on the asyncio driver refactor I've found the emulators to be extremely useful. Especially for the smoothie driver which is so complex.

A big motivator for this PR is to have some incremental PRs from the massive incoming asyncio port PRs.

This PR expands the emulator functionality. 
- The parser can handle long GCODE command strings.
- Most set commands are handles in the modules and smoothie.
- Most importantly the integration tests are expanded.

# Changelog

- Created `Parser` that splits a GCODE line into `Command` objects. 
- Smoothie and module emulators filled out to handle more commands.
- Add `hardware_control.emulation.test_controller` to use emulation for basic testing of HC to smoothie interaction.
- Add `hardware_control.emulation.test_smoothie`. This is a porting of 'drivers.test_driver' to use the smoothie emulator.

# Review requests

I'll try to add comments and questions in PR.

# Risk assessment

No impact except for making tests taking longer to complete. Will be remedied when asyncio drivers are done.

Commit:
expand integration tests.

amitlissack commented on api/src/opentrons/hardware_control/emulation/smoothie.py:
```diff
@@ -1,30 +1,113 @@
+        #  remove 'noqa(C901)'.
+        logger.info(f"Got command {command}")
+        if GCODE.HOMING_STATUS == command.gcode:
+            vals = " ".join(f"{k}:{v}" for k, v in self._home_status.items())
```
Can homing status really just be the axes being at (or near) HOMED_POSITION?

sfoster1 replied:
might be better to emulate the smoothie behavior which is (probably) to keep track of those flags based on parsed homing commands

amitlissack commented on api/src/opentrons/hardware_control/emulation/smoothie.py:
```diff
@@ -1,30 +1,113 @@
+                if 'F' == key:
+                    self._speed = value
+                else:
+                    self._home_status[key] = 0
```
Should a move on an axis clear the homing status of that axis?

Homing status is tricky. As part of the HOME method, the smoothie driver sends a bunch of move commands after the HOME. Specifically on the Y axis.

sfoster1 replied:
the homing status of an axis should be a mapping of axis: bool that is
- initialized to false for all axes
- set to true for an axis when it is homed
- set to false for all axes when the halt line is set

the thing with "homed status" on an open-loop stepper system is that it's not that useful because it can only really be cleared when you know you lost positioning which is pretty much just when you do an emergency stop. You have to be able to move when the axis isn't homed, at least in relative mode, because of the home sequence where we have to move the cross-gantry around to make sure the head is clear of the y homing stop.

amitlissack replied:
Thank you. Exactly what I was looking for.

Commit:
comments and __eq__ in OptionalValue

amitlissack commented on api/tests/opentrons/hardware_control/integration/test_smoothie.py:
```diff
@@ -1,19 +1,441 @@
-    s.connect(f"socket://127.0.0.1:{SMOOTHIE_PORT}")
-    yield s
-    s.disconnect()
+def smoothie(emulation_app) -> driver_3_0.SmoothieDriver_3_0_0:
```
This file is a rebuild of test_driver.py in the driver package.

If we did a char-by-char comparison of expected GCODE here and in test_driver we'd see differences.

The reason is that test_driver takes advantage of `simulation` mode where a function doesn't talk to the smoothie but sets some internal state. 

The values had to change to account for `_setup` and `home` actually sending commands to the smoothie emulator.

SyntaxColoring commented on api/src/opentrons/drivers/types.py:
```diff
@@ -29,6 +29,10 @@ class PlateTemperature(Temperature):
+    UNKNOWN = 'unknown'
+    CLOSED = 'closed'
+    IN_BETWEEN = 'in_between'
+    OPEN = 'open'
+    MAX = 'max'
```
What's the difference between `OPEN` and `MAX`?

amitlissack replied:
I don't know. I got it from the FW. Not sure it's used.

SyntaxColoring commented on api/src/opentrons/hardware_control/emulation/parser.py:
```diff
@@ -0,0 +1,75 @@
+            Command object
+        """
+        previous = None
+        for i in self._re.finditer(line):
```
I haven't tested this, but I *think* using `finditer()` like this will silently discard stuff at the end of the line that we can't parse, and possibly also stuff in the middle of the line that we can't parse.

I think we should error in those cases, instead.

amitlissack replied:
Yes. It only finds matches and ignores the rest.  

I'm going to look into this.

amitlissack replied:
This was a great insight, @SyntaxColoring . Revealed a potential crummy bug.

Commit:
Modify Parser to look for any gcode rather than only a preselected set. This was hiding a bug. If Parser is initialized with G1 as a known gcode and parsed a line like 'G1 G32 F12', the result would be {gcode:G1, params:{G:32, F12}} which is very wrong.
Parser will raise if line starts with non-gcode.

sfoster1 commented on api/src/opentrons/hardware_control/emulation/smoothie.py:
```diff
@@ -1,30 +1,119 @@
+        self._speed = 0.0
+        self._pipette_model = {
+            # p20_multi_v2.0
+            "L": "7032305F6D756C74695F76322E30000000000000000000000000000000000000",
```
might be more readable if encoded here as `extend_to_length(binascii.hexlify(b'p20_multi_v2.0'))` where `extend_to_length` is like
```
def extend_to_length(substr: bytes, length: int = 30):
    return substr + b'0'*(length-len(substr))
```
or whatever

amitlissack replied:
Thanks. DIdn't know about `binascii.hexlify`.

Commit:
remove driver_3_0._byte_array_to_hex_string in favor of utils.string_to_hex.
use string to hex to initialize pipette ids and models in smoothie emulator.
in smoothie emulator, homing status is set to true on home commands and is not reset on movement. once homed the homing statuses will remain true. we need to add a simulation of halt pin to reset these flags.

SyntaxColoring commented on api/src/opentrons/hardware_control/emulation/parser.py:
```diff
@@ -0,0 +1,73 @@
+                    line[previous.end(): i.start()]
+                )
+            else:
+                # This is the first match. It better bet at the beginning or
```
Typo:

```suggestion
                # This is the first match. It better be at the beginning or
```

SyntaxColoring commented on api/src/opentrons/hardware_control/emulation/parser.py:
```diff
@@ -0,0 +1,73 @@
+class Parser:
+    """Gcode line parser."""
+
+    GCODE_RE = re.compile(r"(?:(?:[MG]\d+\.*\d*)|dfu|version)")
```
I think this will incorrectly accept `G123.................456` (with repeated dots). Should this be:
```suggestion
    GCODE_RE = re.compile(r"(?:(?:[MG]\d+\.?\d*)|dfu|version)")
```

SyntaxColoring commented on api/src/opentrons/hardware_control/emulation/parser.py:
```diff
@@ -0,0 +1,73 @@
+
+    GCODE_RE = re.compile(r"(?:(?:[MG]\d+\.*\d*)|dfu|version)")
+    """A gcode is either M or G followed by decimal. Or 'dfu' or 'version'."""
+    PREFIX_NUMBER_RE = re.compile(r"(?P<prefix>[A-Z])(?P<value>-?\d*\.?\d*)")
```
I'd find this clearer if it were named `PREFIXED_NUMBER_RE` (or even `ALPHA_PREFIXED_NUMBER_RE`). `PREFIX_NUMBER_RE` made it initially sound like the number of a prefix (whatever that means), rather than a prefix followed by a number.

Also, for consistency, should the `value` named group instead be named `number`?

amitlissack replied:
ok

SyntaxColoring commented on api/src/opentrons/hardware_control/emulation/parser.py:
```diff
@@ -0,0 +1,73 @@
+                if i.start() != 0:
+                    raise ValueError(f"Invalid content: {line}")
+            previous = i
+        if previous:
+            # Create command from final GCODE and remainder of the line.
+            yield self._create_command(
+                line[previous.start(): previous.end()],
+                line[previous.end():]
+            )
+        elif line:
+            # There are no GCODEs and it's a non-empty line.
+            raise ValueError(f"Invalid content: {line}")
```
If a string starts with valid G-code but has junk at the end, I think this won't detect the error unless the caller exhausts the entire generator.

This probably isn't a big problem, but would it be meaningfully more robust if we checked the whole string for errors before yielding a single command?

For example, could a test erroneously miss an error by naively exhausting only the part of the generator that it cared about, not exhausting it all the way to the end?

amitlissack replied:
If this is a problem in the future we can work on it.

Commit:
little fixes.

mcous commented on api/src/opentrons/drivers/utils.py:
```diff
@@ -145,3 +146,23 @@ def time_since_moved(self) -> Mapping[str, Optional[float]]:
+    hex_string = binascii.hexlify(val.encode()).decode()
+    hex_string_length = len(hex_string)
+    if hex_string_length < min_length:
+        return hex_string + "0" * (min_length - hex_string_length)
```
Forgive me if this is a dumb question, by why is the zero padding on the end rather than the start? Is that how our firmware is encoding the strings into fixed memory? If so, could we mention that the padding is on the end in the docstring?

(I guess that makes sense, start at the start, string ends when you hit null)

amitlissack replied:
I don't know the answer.

amitlissack replied:
Pretty sure the EEPROM stores a 32-byte string that is null terminated. The read commands return the whole thing. I could probably omit the padding.

mcous commented on api/src/opentrons/hardware_control/emulation/connection_handler.py:
```diff
@@ -18,18 +22,15 @@ async def __call__(self, reader: asyncio.StreamReader,
         while True:
```
How does this while loop get broken out of?

amitlissack replied:
It runs forever. Or until the task is canceled.

mcous commented on api/src/opentrons/hardware_control/emulation/magdeck.py:
```diff
@@ -22,25 +28,33 @@ class MagDeckEmulator(AbstractEmulator):
-        self.position = 0
+        self.height: float = 0
+        self.position: float = 0
+        self._parser = Parser()
```
If this `Parser` was passed in as an `__init__` arg, it might make testing this emulator easier. The same would apply if this was a "real" magdeck

mcous commented on api/src/opentrons/hardware_control/emulation/parser.py:
```diff
@@ -0,0 +1,73 @@
+class Parser:
+    """Gcode line parser."""
+
+    GCODE_RE = re.compile(r"(?:(?:[MG]\d+\.?\d*)|dfu|version)")
```
Is zero-prefixed M or G codes an issue for us? It's bitten me before with g-code like languages: `G01 == G1`

amitlissack replied:
What is the issue? Is G01 supposed to be the same as G1?

amitlissack replied:
Added a TODO

mcous commented on api/src/opentrons/hardware_control/emulation/parser.py:
```diff
@@ -0,0 +1,73 @@
+    ALPHA_PREFIXED_NUMBER_RE = re.compile(r"(?P<prefix>[A-Z])(?P<number>-?\d*\.?\d*)")
+    """All parameters are a capital letter followed by a decimal value."""
+
+    def parse(self, line: str) -> Generator[Command, None, None]:
```
[Thoughts for a future PR] In general, I've found this sort of "streaming" parser difficult to work with for g-code-like languages. The ability to look ahead is just a lot less error prone. Given the small size of these strings, I think we'd better off doing something like a full tokenization pass and then a parse on the full tokens list.

That being said, tests on this one look good, so that's totally the sort of optimization we could do later

amitlissack replied:
Makes sense.

mcous commented on api/src/opentrons/hardware_control/emulation/smoothie.py:
```diff
@@ -1,30 +1,114 @@
+            "L": utils.string_to_hex("P3HMV202020041605", 64),
+            "R": utils.string_to_hex("P20SV202020070101", 64),
+        }
+        self._parser = Parser()
```
Same comment about the Parser

mcous reviewed (approved):
A few high-level questions, but tests pass on my machine and this looks great to me!

Commit:
more cleanup.

Commit:
skip the tempdeck temperature integration test because it sometimes fails in CI. this test will soon not require sleeps.
